crates/conductor_lib/src/happ_bundle.rs

```diff
@@ -0,0 +1,189 @@
+use crate::{config::*, port_utils::get_free_port};
+use boolinator::Boolinator;
+use std::collections::HashMap;
+
+#[derive(Serialize, Deserialize)]
+pub struct HappBundle {
+    pub instances: Vec<HappBundleInstance>,
+    pub bridges: Vec<Bridge>,
+    #[serde(rename = "UIs")]
+    pub uis: Vec<HappBundleUi>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct HappBundleInstance {
+    pub name: String,
+    pub id: String,
+    pub dna_hash: String,
+    pub uri: String,
+    pub dna_properties: Option<HashMap<String, String>>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct HappBundleUi {
+    pub name: String,
+    pub id: Option<String>,
+    pub uri: String,
+    pub instance_references: Vec<HappBundleInstanceReference>,
+}
+
+impl HappBundleUi {
+    pub fn id(&self) -> String {
+        self.id.clone().unwrap_or_else(|| String::from(""))
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct HappBundleInstanceReference {
+    pub ui_handle: String,
+    pub instance_id: String,
+}
+
+impl HappBundle {
+    pub fn id_references_are_consistent(&self) -> Result<(), String> {
```

Comment:
This is great. It would be really easy to wasm-bindgen this integrity check code for holoscape so it isn't duped in Rust and Javascript